File: frontend/lib/api/volunteers.ts
import { fetchApi } from './client';

export interface Volunteer {
  id: string;
  name: string;
  latitude: number;
  longitude: number;
  capacity_kg: number;
  has_refrigeration: boolean;
  available_from?: string;
  available_until?: string;
  current_workload_kg: number;
  max_travel_distance_km: number;
  created_at: string;
}

export interface VolunteerCreate {
  id?: string;
  name: string;
  latitude: number;
  longitude: number;
  capacity_kg: number;
  has_refrigeration?: boolean;
  available_from?: string;
  available_until?: string;
  current_workload_kg?: number;
  max_travel_distance_km?: number;
}

export const volunteersApi = {
  list: () => fetchApi<Volunteer[]>('/api/v1/volunteers'),
  get: (id: string) => fetchApi<Volunteer>(`/api/v1/volunteers/${id}`),
  create: (data: VolunteerCreate) => fetchApi<Volunteer>('/api/v1/volunteers', {
    method: 'POST',
    body: JSON.stringify(data),
  }),
};

File: backend/db/repositories/volunteers.py
"""Volunteer repository."""

from typing import List, Optional
from sqlalchemy.orm import Session
from backend.db.models import VolunteerModel
from backend.api.schemas.database import VolunteerCreate


def create_volunteer(db: Session, volunteer: VolunteerCreate) -> VolunteerModel:
    db_volunteer = VolunteerModel(
        name=volunteer.name,
        latitude=volunteer.latitude,
        longitude=volunteer.longitude,
        capacity_kg=volunteer.capacity_kg,
        has_refrigeration=volunteer.has_refrigeration,
        available_from=volunteer.available_from,
        available_until=volunteer.available_until,
        current_workload_kg=volunteer.current_workload_kg,
        max_travel_distance_km=volunteer.max_travel_distance_km,
    )
    db.add(db_volunteer)
    db.commit()
    db.refresh(db_volunteer)
    return db_volunteer


def get_volunteer(db: Session, volunteer_id: str) -> Optional[VolunteerModel]:
    return db.query(VolunteerModel).filter(VolunteerModel.id == volunteer_id).first()


def list_volunteers(db: Session, skip: int = 0, limit: int = 100) -> List[VolunteerModel]:
    return db.query(VolunteerModel).offset(skip).limit(limit).all()


class VolunteerRepository:
    def __init__(self, db: Session):
        self.db = db

    def create(self, volunteer: VolunteerCreate) -> VolunteerModel:
        return create_volunteer(self.db, volunteer)

    def get_by_id(self, volunteer_id: str) -> Optional[VolunteerModel]:
        return get_volunteer(self.db, volunteer_id)

    def list_all(self, skip: int = 0, limit: int = 100) -> List[VolunteerModel]:
        return list_volunteers(self.db, skip=skip, limit=limit)
